src/sing/utility.ts

```diff
@@ -1,3 +1,28 @@
+import { CharacterInfo, StyleInfo, StyleType } from "@/type/preload";
+
+export const isSingingStyle = (styleInfo: StyleInfo) => {
+  return styleInfo.styleType === "humming" || styleInfo.styleType === "sing";
+};
+
+export const filterCharacterInfosByStyleType = (
+  characterInfos: CharacterInfo[],
+  styleType: StyleType | "singerLike"
+) => {
+  const withStylesFiltered = characterInfos.map((characterInfo) => {
+    const styles = characterInfo.metas.styles.filter(
+      styleType === "singerLike"
+        ? isSingingStyle
+        : (styleInfo) => styleInfo.styleType === styleType
+    );
+    return { ...characterInfo, styles };
+  });
+  const withoutEmptyStyles = withStylesFiltered.filter(
+    (characterInfo) => characterInfo.styles.length > 0
+  );
```

Comment:
```suggestion
  const withStylesFiltered = characterInfos.map((characterInfo) => {
    const styles = characterInfo.metas.styles.filter(
      styleType === "singerLike"
        ? isSingingStyle
        : (styleInfo) => styleInfo.styleType === styleType
    );
    return {
      ...characterInfo,
      metas: {
        ...characterInfo.metas,
        styles,
      },
    };
  });
  const withoutEmptyStyles = withStylesFiltered.filter(
    (characterInfo) => characterInfo.metas.styles.length > 0
  );
```

こうしないとobjectの型が変わってしまっていると思います...!(stylesのフィルタリングもうまくできていなさそう)

あと、`styleType`の判定`styleInfo.styleType === styleType`は`talk`タイプの時に機能しないため(マルチエンジンにおいては、`undefined`になりうるため)、`USER_ORDERED_CHARACTER_INFOS`と同じロジックにしないと、表示されないキャラクターが現れてしまいます。